platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/migration/CipherUtils.java

```diff
@@ -0,0 +1,205 @@
+      return iCipher;
+    } catch (InvalidKeyException
+        | NoSuchPaddingException
+        | NoSuchAlgorithmException
+        | InvalidAlgorithmParameterException e) {
+      throw new MigrationException(MigrationZipConstants.KEY_INVALID_ERROR, keyPath, e);
+    }
+  }
+
+  /**
+   * Creates a file containing the checksum for the zip file
+   *
+   * @throws IOException
+   * @throws NoSuchAlgorithmException
+   */
+  public void createZipChecksumFile() throws IOException, NoSuchAlgorithmException {
+    String hash = getChecksum(zipPath);
+    writeStringToFile(checksumPath.toFile(), hash, Charsets.UTF_8);
+  }
+
+  @SuppressWarnings({ //
+    "squid:S2093", /* try-with-resource would require that the input stream is initialized too early */
+    "squid:S2095" /* stream is closed in finally clause */
+  })
+  private SecretKey initKey(Path keyPath) throws NoSuchAlgorithmException {
+    if (keyPath.toFile().exists()) {
+      SecretKey key;
+      try {
+        byte[] keyBytes = decodeHex(loadKeyDataFrom(keyPath).toCharArray());
+        key = new SecretKeySpec(keyBytes, MigrationZipConstants.KEY_ALGORITHM);
+      } catch (DecoderException e) {
+        String message = String.format(MigrationZipConstants.KEY_DECODE_ERROR, keyPath, e);
+        LOGGER.warn(message);
+        throw new MigrationException(message);
+      }
+      this.secretKey = key;
+      return key;
+    } else {
+      return createSecretKey(keyPath);
+    }
+  }
+
+  private String loadKeyDataFrom(Path keyPath) {
+    try (BufferedInputStream bif = new BufferedInputStream(new FileInputStream(keyPath.toFile()))) {
+      return IOUtils.toString(bif, StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      String message = String.format(MigrationZipConstants.FILE_IO_ERROR, keyPath, e);
+      LOGGER.warn(message);
+      throw new MigrationException(message);
+    }
+  }
+
+  /**
+   * Creates a secret key and stores it in a file
+   *
+   * @param keyPath the path for storing the secret key
+   * @throws MigrationException when an invalid key algorithm is used
+   * @throws MigrationException when the key can not be written to a file
+   * @return a {@link SecretKey}
+   */
+  private SecretKey createSecretKey(Path keyPath)
+      throws MigrationException, NoSuchAlgorithmException {
+    KeyGenerator keyGenerator = null;
+    try {
+      keyGenerator = KeyGenerator.getInstance(MigrationZipConstants.KEY_ALGORITHM);
+      secretKey = keyGenerator.generateKey();
+      char[] hexKey = encodeHex(secretKey.getEncoded());
+      writeStringToFile(keyPath.toFile(), String.valueOf(hexKey), Charsets.UTF_8);
+      return secretKey;
+    } catch (IOException e) {
+      throw new MigrationException(String.format(MigrationZipConstants.FILE_IO_ERROR, keyPath, e));
+    }
+  }
+
+  private String getChecksum(Path path) throws IOException, NoSuchAlgorithmException {
+    try (FileInputStream fis = new FileInputStream(path.toFile())) {
```

Comment:
✏️ This one, on the other hand, is going to be pretty big and buffering it might be of value.

Reply:
Added